tests/scripts/all.sh

```diff
@@ -2110,6 +2110,36 @@ component_test_psa_crypto_config_accel_hash_use_psa () {
     fi
 }
 
+component_test_psa_crypto_config_reference_hash_use_psa() {
```

Comment:
There should be a prominent comment explaining the purpose of this component and that in need to be kept in sync with `accel_hash_use_psa()` and the data in `analyze_outcomes.py`. We also need a comment in `accel_hash_use_psa()` about keeping things in sync.

Reply:
Shouldn't we aim to keep these configs as close as possible?
I can see that there's a number of differences between these two, and we opt to ignore test suites instead:
```
    scripts/config.py set MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG
    scripts/config.py unset MBEDTLS_ENTROPY_C
    scripts/config.py unset MBEDTLS_ENTROPY_NV_SEED # depends on ENTROPY_C
    scripts/config.py unset MBEDTLS_PLATFORM_NV_SEED_ALT # depends on former
    # Also unset MD_C and things that depend on it;
    # see component_test_crypto_full_no_md.
    scripts/config.py unset MBEDTLS_MD_C
    scripts/config.py unset MBEDTLS_HKDF_C # has independent PSA implementation
    scripts/config.py unset MBEDTLS_HMAC_DRBG_C
    scripts/config.py unset MBEDTLS_ECDSA_DETERMINISTIC
    ```
Can we get rid of any of those?

Reply:
This is a configuration without `MBEDTLS_MD_C`, and what you listed here is all the things that depend on `MBEDTLS_MD_C` plus disabling entropy since the PSA fake RNG is in use.

Reply:
I don't mean enabling any of these in a build without `MBEDTLS_MD_C` (`component_test_psa_crypto_config_accel_hash_use_psa`), I mean disabling as many of these as possible in `component_test_psa_crypto_config_reference_hash_use_psa` to get closer to the configuration in the first mentioned component.

Reply:
`MBEDTLS_ECDSA_DETERMINISTIC` depends on `MBEDTLS_HMAC_DRBG_C`, and `MBEDTLS_HMAC_DRBG_C` on `MBEDTLS_MD_C`. 
`MBEDTLS_ECDSA_DETERMINISTIC` is disabled as nothing else depends on it.
If `MBEDTLS_PSA_CRYPTO_C` and `MBEDTLS_ENTROPY_C` is defined, either of  `defined(MBEDTLS_CTR_DRBG_C) || defined(MBEDTLS_HMAC_DRBG_C) ` have to also be defined.
To get as close as possible to `component_test_psa_crypto_config_accel_hash_use_psa`, we could add `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` and disable `MBEDTLS_ENTROPY_C`, but maybe that's going too far.
Less invasive approach would be to use `MBEDTLS_ENTROPY_C` and `MBEDTLS_CTR_DRBG_C` and disable `MBEDTLS_HMAC_DRBG_C`.

Reply:
> I mean disabling as many of these as possible in `component_test_psa_crypto_config_reference_hash_use_psa` to get closer to the configuration in the first mentioned component.

I'd like to ask: why do we want to get closer to the configuration in the driver component? So far I've worked on the (implicit) that we only need to make it close enough for the comparison to find no difference once we exclude the right suites. But indeed we could take the opposite approach: make the reference configuration very close, and then in the comparison script the ignore list would become much smaller, ideally just mdx, shax, md here (as Gilles said, until we change things regarding MD, but that's actually quite orthogonal to the present discussion I think).

So, two possible approaches here. Now that you mention it, I think I like the idea of making the two configurations (reference and drivers) very close, so that we only need a very small ignore list in the comparison script. It's a bit awkward that we currently have to ignore `test_suite_psa_crypto_init` for example, and I think this would go away if we also used `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` in the reference config indeed.

@mprse can you try that out and see how it works?

Reply:
Yes I can check this, but we need to define how the reference implementation should be constructed.

```
scripts/config.py full
# use PSA config and disable driver-less algs as in the component
scripts/config.py set MBEDTLS_PSA_CRYPTO_CONFIG
scripts/config.py -f include/psa/crypto_config.h unset PSA_WANT_ALG_STREAM_CIPHER
scripts/config.py -f include/psa/crypto_config.h unset PSA_WANT_ALG_ECB_NO_PADDING
# Disable same modules as in accelerated configuration and Use an external RNG.
scripts/config.py set MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG
scripts/config.py unset MBEDTLS_ENTROPY_C
scripts/config.py unset MBEDTLS_ENTROPY_NV_SEED # depends on ENTROPY_C
scripts/config.py unset MBEDTLS_PLATFORM_NV_SEED_ALT # depends on former
scripts/config.py unset MBEDTLS_HKDF_C # has independent PSA implementation
scripts/config.py unset MBEDTLS_HMAC_DRBG_C
scripts/config.py unset MBEDTLS_ECDSA_DETERMINISTIC
scripts/config.py -f include/psa/crypto_config.h unset PSA_WANT_ALG_DETERMINISTIC_ECDSA
```

Reply:
Yes, that looks right. Actually what we want, using Andrzej's approach, is basically a copy of [this block](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/all.sh#L2080-L2108) except we skip [this sub-block](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/all.sh#L2085-L2094) and [this line](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/all.sh#L2104) - we could even keep them but commented out to make the similarity (and differences) more apparent. Or we could even strive to avoid the duplication and extract that to an auxiliary function that takes a parameter indicated whether to keep software hashes or not:
```sh
config_psa_crypto_hash_use_psa() {
    DRIVER_ONLY="$1"
    # start with config full for maximum coverage (also enables USE_PSA)
    scripts/config.py full
    # enable support for drivers and configuring PSA-only algorithms
    scripts/config.py set MBEDTLS_PSA_CRYPTO_DRIVERS
    scripts/config.py set MBEDTLS_PSA_CRYPTO_CONFIG
    if [ "$DRIVER_ONLY" -eq 1 ]; then
        # disable the built-in implementation of hashes
        scripts/config.py unset MBEDTLS_MD5_C
        scripts/config.py unset MBEDTLS_RIPEMD160_C
        scripts/config.py unset MBEDTLS_SHA1_C
        scripts/config.py unset MBEDTLS_SHA224_C
        scripts/config.py unset MBEDTLS_SHA256_C # see external RNG below
        scripts/config.py unset MBEDTLS_SHA256_USE_A64_CRYPTO_IF_PRESENT
        scripts/config.py unset MBEDTLS_SHA384_C
        scripts/config.py unset MBEDTLS_SHA512_C
        scripts/config.py unset MBEDTLS_SHA512_USE_A64_CRYPTO_IF_PRESENT
    if
    # Use an external RNG as currently internal RNGs depend on entropy.c
    # which in turn hard-depends on SHA256_C (or SHA512_C).
    # See component_test_psa_external_rng_no_drbg_use_psa.
    scripts/config.py set MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG
    scripts/config.py unset MBEDTLS_ENTROPY_C
    scripts/config.py unset MBEDTLS_ENTROPY_NV_SEED # depends on ENTROPY_C
    scripts/config.py unset MBEDTLS_PLATFORM_NV_SEED_ALT # depends on former
    # Also unset MD_C and things that depend on it;
    # see component_test_crypto_full_no_md.
    if [ "$DRIVER_ONLY" -eq 1 ]; then
        scripts/config.py unset MBEDTLS_MD_C
    fi
    scripts/config.py unset MBEDTLS_HKDF_C # has independent PSA implementation
    scripts/config.py unset MBEDTLS_HMAC_DRBG_C
    scripts/config.py unset MBEDTLS_ECDSA_DETERMINISTIC
    scripts/config.py -f include/psa/crypto_config.h unset PSA_WANT_ALG_DETERMINISTIC_ECDSA
```
This would also avoid synchronization issue when we change one of the components. Wdyt?

Reply:
> Yes, that looks right. Actually what we want, using Andrzej's approach, is basically a copy of [this block](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/all.sh#L2080-L2108) except we skip [this sub-block](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/all.sh#L2085-L2094) and [this line](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/all.sh#L2104)

Yes this is what I did making this proposition. I like the proposition with parameter, but it seems that then we will have the same component name for reference and drivers in outcome file.

Reply:
I don't think so: the auxiliary function would be used by both components, but each component would still have its own `component_xxx()` function (note the auxiliary function is only doing the config, not the build or testing) giving the correct name in the outcome file.

Reply:
`scripts/config.py set MBEDTLS_PSA_CRYPTO_DRIVERS` should be also under `if [ "$DRIVER_ONLY" -eq 1 ]; then`.

Reply:
Good point, I failed to consider that. Though I'm not sure, actually, does it hurt to have support for drivers in the reference config, even if we still use the built-in implementation?

Reply:
Hmm. From my perspective in this case built-in implementation should be used even if drivers are enabled (without accelerated hash), but it might be misleading.

Reply:
Yes, I agree, so I think we don't need to make `scripts/config.py set MBEDTLS_PSA_CRYPTO_DRIVERS` conditional - it can be in both reference and driver-based configs.

Reply:
Tests were already running with disabled `MBEDTLS_PSA_CRYPTO_DRIVERS` and because it takes some time I finished run and got the following results:

```
# Auxiliary function to build config with build-in hashes and accelerated hashes
config_psa_crypto_hash_use_psa () {
    DRIVER_ONLY="$1"
    # start with config full for maximum coverage (also enables USE_PSA)
    scripts/config.py full
    # enable support for configuring PSA-only algorithms
    scripts/config.py set MBEDTLS_PSA_CRYPTO_CONFIG
    if [ "$DRIVER_ONLY" -eq 1 ]; then
        # enable support for drivers
        scripts/config.py set MBEDTLS_PSA_CRYPTO_DRIVERS
        # disable the built-in implementation of hashes
        scripts/config.py unset MBEDTLS_MD5_C
        scripts/config.py unset MBEDTLS_RIPEMD160_C
        scripts/config.py unset MBEDTLS_SHA1_C
        scripts/config.py unset MBEDTLS_SHA224_C
        scripts/config.py unset MBEDTLS_SHA256_C # see external RNG below
        scripts/config.py unset MBEDTLS_SHA256_USE_A64_CRYPTO_IF_PRESENT
        scripts/config.py unset MBEDTLS_SHA384_C
        scripts/config.py unset MBEDTLS_SHA512_C
        scripts/config.py unset MBEDTLS_SHA512_USE_A64_CRYPTO_IF_PRESENT
    fi
    # Use an external RNG as currently internal RNGs depend on entropy.c
    # which in turn hard-depends on SHA256_C (or SHA512_C).
    # See component_test_psa_external_rng_no_drbg_use_psa.
    scripts/config.py set MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG
    scripts/config.py unset MBEDTLS_ENTROPY_C
    scripts/config.py unset MBEDTLS_ENTROPY_NV_SEED # depends on ENTROPY_C
    scripts/config.py unset MBEDTLS_PLATFORM_NV_SEED_ALT # depends on former
    # Also unset MD_C and things that depend on it;
    # see component_test_crypto_full_no_md.
    if [ "$DRIVER_ONLY" -eq 1 ]; then
        scripts/config.py unset MBEDTLS_MD_C
    fi
    scripts/config.py unset MBEDTLS_HKDF_C # has independent PSA implementation
    scripts/config.py unset MBEDTLS_HMAC_DRBG_C
    scripts/config.py unset MBEDTLS_ECDSA_DETERMINISTIC
    scripts/config.py -f include/psa/crypto_config.h unset PSA_WANT_ALG_DETERMINISTIC_ECDSA
}

# Note that component_test_psa_crypto_config_reference_hash_use_psa
# is related to this component and both components need to be kept in sync.
# For details please see comments for component_test_psa_crypto_config_reference_hash_use_psa.
component_test_psa_crypto_config_accel_hash_use_psa () {
    msg "test: MBEDTLS_PSA_CRYPTO_CONFIG with accelerated hash and USE_PSA"

    # Disable ALG_STREAM_CIPHER and ALG_ECB_NO_PADDING to avoid having
    # partial support for cipher operations in the driver test library.
    scripts/config.py -f include/psa/crypto_config.h unset PSA_WANT_ALG_STREAM_CIPHER
    scripts/config.py -f include/psa/crypto_config.h unset PSA_WANT_ALG_ECB_NO_PADDING

    loc_accel_list="ALG_MD5 ALG_RIPEMD160 ALG_SHA_1 ALG_SHA_224 ALG_SHA_256 ALG_SHA_384 ALG_SHA_512"
    loc_accel_flags=$( echo "$loc_accel_list" | sed 's/[^ ]* */-DLIBTESTDRIVER1_MBEDTLS_PSA_ACCEL_&/g' )
    make -C tests libtestdriver1.a CFLAGS="$ASAN_CFLAGS $loc_accel_flags" LDFLAGS="$ASAN_CFLAGS"

    config_psa_crypto_hash_use_psa 1

    loc_accel_flags="$loc_accel_flags $( echo "$loc_accel_list" | sed 's/[^ ]* */-DMBEDTLS_PSA_ACCEL_&/g' )"
    make CFLAGS="$ASAN_CFLAGS -Werror -I../tests/include -I../tests -I../../tests -DPSA_CRYPTO_DRIVER_TEST -DMBEDTLS_TEST_LIBTESTDRIVER1 $loc_accel_flags" LDFLAGS="-ltestdriver1 $ASAN_CFLAGS" all

    # There's a risk of something getting re-enabled via config_psa.h;
    # make sure it did not happen.
    not grep mbedtls_md library/md.o
    not grep mbedtls_md5 library/md5.o
    not grep mbedtls_sha1 library/sha1.o
    not grep mbedtls_sha256 library/sha256.o
    not grep mbedtls_sha512 library/sha512.o
    not grep mbedtls_ripemd160 library/ripemd160.o

    msg "test: MBEDTLS_PSA_CRYPTO_CONFIG with accelerated hash and USE_PSA"
    make test

    msg "test: ssl-opt.sh, MBEDTLS_PSA_CRYPTO_CONFIG with accelerated hash and USE_PSA"
    tests/ssl-opt.sh

    msg "test: compat.sh, MBEDTLS_PSA_CRYPTO_CONFIG without accelerated hash and USE_PSA"
    #tests/compat.sh
}

# This component provides reference configuration for test_psa_crypto_config_accel_hash_use_psa
# without accelerated hash. The outcome from both components are used by the analyze_outcomes.py
# script to find regression in test coverage when accelerated hash is used (tests and ssl-opt).
# Both components need to be kept in sync.
component_test_psa_crypto_config_reference_hash_use_psa() {
    msg "test: MBEDTLS_PSA_CRYPTO_CONFIG without accelerated hash and USE_PSA"

    config_psa_crypto_hash_use_psa 0

    make

    msg "test: MBEDTLS_PSA_CRYPTO_CONFIG without accelerated hash and USE_PSA"
    make test

    msg "test: ssl-opt.sh, MBEDTLS_PSA_CRYPTO_CONFIG without accelerated hash and USE_PSA"
    tests/ssl-opt.sh
}
```

But unfortunately the results are different (I didn't modify skipped tests in python script):
```
przemek@przemek-VirtualBox:~/mbedtls$ tests/scripts/analyze_outcomes.py outcome.csv analyze_driver_vs_reference_hash
test_suite_psa_crypto;PSA AEAD setup: invalid algorithm (ChaCha20): driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA cipher decrypt: invalid key type: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA cipher decrypt: without initialization: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA cipher encrypt: invalid key type: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA cipher encrypt: without initialization: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA key policy: AES ECB: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt multipart: 2-key 3DES-ECB, 8 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt multipart: 3-key 3DES-ECB, 8 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt multipart: AES-ECB, 0 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt multipart: AES-ECB, 16 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt multipart: AES-ECB, 32 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt multipart: AES-ECB, input too short (15 bytes): driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt multipart: ChaCha20, K=rand N=rand: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt: 2-key 3DES-ECB, 8 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt: 3-key 3DES-ECB, 8 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt: AES-ECB, 0 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt: AES-ECB, 16 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt: AES-ECB, 32 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt: AES-ECB, input too short (15 bytes): driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt: ChaCha20, RFC7539 keystream: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decrypt: ChaCha20, RFC7539 sunscreen: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decryption multipart: AES-ECB, 11+21 bytes: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decryption multipart: AES-ECB, 16+16 bytes: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decryption multipart: AES-ECB, 28+4 bytes: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric decryption multipart: ChaCha20, 14+50 bytes: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt multipart: 2-key 3DES-ECB, 8 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt multipart: 3-key 3DES-ECB, 8 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt multipart: AES-ECB, 0 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt multipart: AES-ECB, 16 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt multipart: AES-ECB, 32 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt multipart: AES-ECB, input too short (15 bytes): driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt multipart: ChaCha20, K=0 N=0: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt multipart: ChaCha20, K=rand N=rand: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt validation: ChaCha20, K=0 N=0: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt validation: ChaCha20, K=rand N=rand: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt/decrypt multipart: AES-ECB, 16 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt/decrypt: AES-ECB, 16 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt: 2-key 3DES-ECB, 8 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt: 3-key 3DES-ECB, 8 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt: AES-ECB, 0 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt: AES-ECB, 16 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt: AES-ECB, 32 bytes, good: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encrypt: AES-ECB, input too short (15 bytes): driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encryption multipart: AES-ECB, 13+19 bytes: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encryption multipart: AES-ECB, 16+16 bytes: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encryption multipart: AES-ECB, 24+12 bytes: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encryption multipart: ChaCha20, 14+50 bytes: driver: skipped/failed; reference: passed
test_suite_psa_crypto;PSA symmetric encryption multipart: ChaCha20, RFC7539 sunscreen: driver: skipped/failed; reference: passed
test_suite_psa_crypto_metadata;Cipher: ECB-nopad: driver: skipped/failed; reference: passed
test_suite_psa_crypto_metadata;Cipher: STREAM_CIPHER: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA aead ECB_NO_PADDING: invalid with AES: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA aead ECB_NO_PADDING: invalid with ARIA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA aead ECB_NO_PADDING: invalid with CAMELLIA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA aead ECB_NO_PADDING: invalid with DES: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA aead STREAM_CIPHER: invalid with CHACHA20: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA asymmetric_encryption ECB_NO_PADDING: invalid with AES: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA asymmetric_encryption ECB_NO_PADDING: invalid with ARIA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA asymmetric_encryption ECB_NO_PADDING: invalid with CAMELLIA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA asymmetric_encryption ECB_NO_PADDING: invalid with DES: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA asymmetric_encryption STREAM_CIPHER: invalid with CHACHA20: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher ECB_NO_PADDING: incompatible with CHACHA20: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher ECB_NO_PADDING: incompatible with DERIVE: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher ECB_NO_PADDING: incompatible with HMAC: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher ECB_NO_PADDING: incompatible with RAW_DATA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher ECB_NO_PADDING: incompatible with RSA_KEY_PAIR: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher ECB_NO_PADDING: incompatible with RSA_PUBLIC_KEY: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher STREAM_CIPHER: incompatible with AES: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher STREAM_CIPHER: incompatible with ARIA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher STREAM_CIPHER: incompatible with CAMELLIA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher STREAM_CIPHER: incompatible with DERIVE: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher STREAM_CIPHER: incompatible with DES: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher STREAM_CIPHER: incompatible with HMAC: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher STREAM_CIPHER: incompatible with RAW_DATA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher STREAM_CIPHER: incompatible with RSA_KEY_PAIR: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA cipher STREAM_CIPHER: incompatible with RSA_PUBLIC_KEY: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA hash ECB_NO_PADDING: invalid: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA hash STREAM_CIPHER: invalid: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA key_agreement ECB_NO_PADDING: invalid with AES: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA key_agreement ECB_NO_PADDING: invalid with ARIA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA key_agreement ECB_NO_PADDING: invalid with CAMELLIA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA key_agreement ECB_NO_PADDING: invalid with DES: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA key_agreement STREAM_CIPHER: invalid with CHACHA20: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA key_derivation ECB_NO_PADDING: invalid: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA key_derivation STREAM_CIPHER: invalid: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA mac ECB_NO_PADDING: invalid with AES: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA mac ECB_NO_PADDING: invalid with ARIA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA mac ECB_NO_PADDING: invalid with CAMELLIA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA mac ECB_NO_PADDING: invalid with DES: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA mac STREAM_CIPHER: invalid with CHACHA20: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA sign ECB_NO_PADDING: invalid with AES: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA sign ECB_NO_PADDING: invalid with ARIA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA sign ECB_NO_PADDING: invalid with CAMELLIA: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA sign ECB_NO_PADDING: invalid with DES: driver: skipped/failed; reference: passed
test_suite_psa_crypto_op_fail.generated;PSA sign STREAM_CIPHER: invalid with CHACHA20: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save alg2: ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save alg2: STREAM_CIPHER: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save alg: ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save alg: STREAM_CIPHER: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save type: AES 128-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save type: AES 192-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save type: AES 256-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save type: ARIA 128-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save type: ARIA 192-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save type: ARIA 256-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save type: CAMELLIA 128-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save type: CAMELLIA 192-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save type: CAMELLIA 256-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save type: CHACHA20 256-bit, STREAM_CIPHER: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save type: DES 128-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save type: DES 192-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.current;save type: DES 64-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read alg2: ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read alg2: STREAM_CIPHER: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read alg: ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read alg: STREAM_CIPHER: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read type: AES 128-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read type: AES 192-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read type: AES 256-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read type: ARIA 128-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read type: ARIA 192-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read type: ARIA 256-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read type: CAMELLIA 128-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read type: CAMELLIA 192-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read type: CAMELLIA 256-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read type: CHACHA20 256-bit, STREAM_CIPHER: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read type: DES 128-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read type: DES 192-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
test_suite_psa_crypto_storage_format.v0;read type: DES 64-bit, ECB_NO_PADDING: driver: skipped/failed; reference: passed
```

Additional analysis is required.
I will look into this after creating patch for j-pake password.

Reply:
It looks like the difference in test coverage is caused by:
- PSA_WANT_ALG_STREAM_CIPHER
- PSA_WANT_ALG_ECB_NO_PADDING

`include/psa/crypto_config.h` diff for driver vs reference:
![image](https://user-images.githubusercontent.com/30721012/202139168-94406f98-7fd7-436a-b0d4-d6044908e71e.png)

This is because:
```
    # Disable ALG_STREAM_CIPHER and ALG_ECB_NO_PADDING to avoid having
    # partial support for cipher operations in the driver test library.
    scripts/config.py -f include/psa/crypto_config.h unset PSA_WANT_ALG_STREAM_CIPHER
    scripts/config.py -f include/psa/crypto_config.h unset PSA_WANT_ALG_ECB_NO_PADDING
```

But why this was not the case with previous configurations?

Reply:
Disabling ECB and STREAM_CIPHER was done because of limitations of driver dispatch and/or test drivers at the time. It might not be necessary anymore.

Reply:
Results are fine with disabled `PSA_WANT_ALG_STREAM_CIPHER, PSA_WANT_ALG_ECB_NO_PADDING` also in reference config and reduced skipped tests in script. Please review.